### PR TITLE
Add validation 

### PR DESCRIPTION
**Pull request by Sushmithamallesh**
Issue: #22

### CODE REVIEW

**Comment by Sushmithamallesh** [comment]
I will raise a separate PR for using the validation error for retry since that needs to done for the simple case of type mismatch too.

More importantly I didn't understand the necessity for `WithValidator`. Feel free to point out if I am missing something @h0rv

**Comment by h0rv** [comment]
Hi @Sushmithamallesh, we have 2 choices for this implementation.

So, the `validate` struct tags are not required currently, and maybe shouldn't be required at all, but should be an opt in functionality for the package.

We can either

1. use `reflect` package to check if the validate flags exist on the struct passed to the library, or
2. Use a flag like `WithValidator` that the user has to explicitly pass for us to do the validation logic on our end.

If we do the first approach, we should add a flag then to turn off validation. The validation code will add overhead and should let the users opt out if performance matters.

With either approach, we should conditionally run the validation logic on the returned responses.

Thanks for the PR and let me know if you have any other questions!

**Comment by Sushmithamallesh** [comment]
Oh right. I should have considered the performance that is unnecessary overhead for sure. The second option does sound a lot more better, since it's an opt in. 
Will make changes to the PR to reflect that.

**Comment by Sushmithamallesh** [comment]
@h0rv Have made a bunch of changes. Please take a look. Thanks!

File: examples/validator/main.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	fmt.Printf("First Name: %s\n", user.FirstName)
+	fmt.Printf("Last Name: %s\n", user.LastName)
+	fmt.Printf("Age: %d\n", user.Age)
+	fmt.Printf("Email: %s\n", user.Email)
+	fmt.Printf("Gender: %s\n", user.Gender)
+	fmt.Printf("Favourite Color: %s\n", user.FavouriteColor)
+	fmt.Println("Addresses:")
+	for _, address := range user.Addresses {
+		fmt.Printf("\tStreet: %s, City: %s, Planet: %s, Phone: %s\n", address.Street, address.City, address.Planet, address.Phone)
+	}
```
Comment (h0rv) [inline examples/validator/main.go:70]: put expected comment in block comment below

File: examples/validator/main.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/instructor-ai/instructor-go/pkg/instructor"
+	openai "github.com/sashabaranov/go-openai"
+)
+
+// User contains user information
+type User struct {
```
Comment (h0rv) [inline examples/validator/main.go:14]: structs still need `json` and `jsonschema` tags. Curious why this still works without `jsonschema`

File: pkg/instructor/anthropic.go [hunk 3]
```diff
@@ -7,9 +7,10 @@ import (
 type InstructorAnthropic struct {
 	*anthropic.Client
 
-	provider   Provider
-	mode       Mode
-	maxRetries int
+	provider      Provider
+	mode          Mode
+	maxRetries    int
+	withValidator bool
```
Comment (h0rv) [inline pkg/instructor/anthropic.go:13]: `withValidator` --> `validate`

File: pkg/instructor/anthropic.go [hunk 4]
```diff
@@ -21,9 +22,10 @@ func FromAnthropic(client *anthropic.Client, opts ...Options) *InstructorAnthrop
 	i := &InstructorAnthropic{
 		Client: client,
 
-		provider:   ProviderOpenAI,
-		mode:       *options.Mode,
-		maxRetries: *options.MaxRetries,
+		provider:      ProviderOpenAI,
+		mode:          *options.Mode,
+		maxRetries:    *options.MaxRetries,
+		withValidator: *options.WithValidator,
```
Comment (h0rv) [inline pkg/instructor/anthropic.go:28]: `withValidator` --> `validate`

File: pkg/instructor/chat.go [hunk 5]
```diff
@@ -5,12 +5,16 @@ import (
...
 	"reflect"
+
+	"github.com/go-playground/validator/v10"
 )
 
 func chatHandler(i Instructor, ctx context.Context, request interface{}, response any) (interface{}, error) {
 
 	var err error
 
+	validate = validator.New()
```
Comment (h0rv) [inline pkg/instructor/chat.go:16]: only create the validator if `i.validate` is set true

File: pkg/instructor/chat_stream.go [hunk 6]
```diff
@@ -15,6 +17,8 @@ const WRAPPER_END = `"items": [`
 
 func chatStreamHandler(i Instructor, ctx context.Context, request interface{}, response any) (<-chan interface{}, error) {
 
+	validate = validator.New()
```
Comment (h0rv) [inline pkg/instructor/chat_stream.go:20]: only create validator if `i.validate` is true

File: pkg/instructor/instructor.go [hunk 7]
```diff
@@ -2,12 +2,17 @@ package instructor
...
+	"github.com/go-playground/validator/v10"
 )
 
+var validate *validator.Validate
+
 type Instructor interface {
 	Provider() Provider
 	Mode() Mode
 	MaxRetries() int
+	WithValidator() bool
```
Comment (h0rv) [inline pkg/instructor/instructor.go:15]: `WithValidator()` --> `Validate()`

File: pkg/instructor/cohere_struct.go [hunk 8]
```diff
@@ -7,9 +7,10 @@ import (
 type InstructorCohere struct {
 	*cohere.Client
 
-	provider   Provider
-	mode       Mode
-	maxRetries int
+	provider      Provider
+	mode          Mode
+	maxRetries    int
+	withValidator bool
```
Comment (h0rv) [inline pkg/instructor/cohere_struct.go:13]: `withValidator` --> `validate`

File: pkg/instructor/options.go [hunk 9]
```diff
@@ -24,13 +26,20 @@ func WithMaxRetries(maxRetries int) Options {
 	return Options{MaxRetries: toPtr(maxRetries)}
 }
 
+func WithValidator(withValidator bool) Options {
```
Comment (h0rv) [inline pkg/instructor/options.go:29]: we can either use this name or `WithValidation`

**Comment by h0rv** [comment]
Hi @Sushmithamallesh, I left some comments, but looks good!

**Comment by h0rv** [comment]
I also just pushed some changes to the example to better showcase this feature

File: pkg/instructor/anthropic.go (hunk 3 above)
Comment (Sushmithamallesh) [reply pkg/instructor/anthropic.go]: done

File: pkg/instructor/anthropic.go (hunk 4 above)
Comment (Sushmithamallesh) [reply pkg/instructor/anthropic.go]: done

File: pkg/instructor/cohere_struct.go (hunk 8 above)
Comment (Sushmithamallesh) [reply pkg/instructor/cohere_struct.go]: done

File: pkg/instructor/options.go [hunk 10]
```diff
@@ -2,18 +2,20 @@ package instructor
...
+	DefaultValidator  = false
 )
 
 type Options struct {
-	Mode       *Mode
-	MaxRetries *int
-
+	Mode          *Mode
+	MaxRetries    *int
+	WithValidator *bool
```
[the remark this answers is not in the document]
Comment (Sushmithamallesh) [reply pkg/instructor/options.go]: done

File: pkg/instructor/options.go (hunk 9 above)
Comment (Sushmithamallesh) [reply pkg/instructor/options.go]: changed it to WithValidation. sounds better!

[the remark this answers is not in the document]
Comment (Sushmithamallesh) [reply pkg/instructor/options.go]: done

**Review by h0rv — APPROVED** [review]
LGTM. Thanks for the feature - much appreciated!

**Pull request merged**